akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala

```diff
@@ -23,59 +23,70 @@ import scala.concurrent.Future
-  def forwardToUri(uri: Uri, circuitBreakerOption: Option[CircuitBreaker] = None)(implicit system: ActorSystem): Directive1[Try[HttpResponse]] = {
-    extractMaterializer.flatMap { implicit mat ⇒
-      val requestUriInterceptor = ProxyUriRequestInterceptor(uri).andThen(ProxyHeaderRequestInterceptor)
-      //TODO: select executor from configuration
-      val executor = CachedHostConnectionPoolRequestExecutor()
-      forwardRequest(executor, requestUriInterceptor, circuitBreakerOption)
+  def forwardToUri(uri: Uri, configOption: Option[ReverseProxyConfig] = None): Directive1[Try[HttpResponse]] = {
+    val requestUriInterceptor = ProxyUriRequestInterceptor(uri).andThen(ProxyHeaderRequestInterceptor)
+    mapRequest(requestUriInterceptor).tflatMap { _ ⇒
+      forwardRequest(configOption)
     }
   }
 
-  def forwardRequest(requestExecutor: RequestExecutor, requestInterceptor: RequestInterceptor, circuitBreakerOption: Option[CircuitBreaker] = None): Directive1[Try[HttpResponse]] = {
+  def forwardRequest(configOption: Option[ReverseProxyConfig] = None): Directive1[Try[HttpResponse]] = {
     extractExecutionContext.flatMap { implicit ec ⇒
-      extractMaterializer.flatMap { implicit mat ⇒
-        extractRequest.flatMap { request ⇒
-          val interceptedRequest: HttpRequest = requestInterceptor(request)
-          val responseFuture: Future[HttpResponse] = requestExecutor(interceptedRequest)
-          circuitBreakerOption.fold(onComplete(responseFuture))(circuitBreaker ⇒ onCompleteWithBreaker(circuitBreaker)(responseFuture))
+      extractRequest.flatMap { request ⇒
+        getOrCreateDefaultConfig(configOption).flatMap { reverseProxyConfig ⇒
+          val responseFuture: Future[HttpResponse] = reverseProxyConfig.requestExecutor.apply(request)
+          reverseProxyConfig.circuitBreakerOption match {
+            case Some(circuitBreaker) ⇒ onCompleteWithBreaker(circuitBreaker)(responseFuture)
+            case None                 ⇒ onComplete(responseFuture)
+          }
+        }
+      }
+    }
+  }
+
+  private[akka] def getOrCreateDefaultConfig(configOption: Option[ReverseProxyConfig]): Directive1[ReverseProxyConfig] = {
+    configOption match {
+      case Some(config) ⇒ provide(config)
+      case None ⇒ {
+        extractMaterializer.flatMap { mat ⇒
+          extractActorSystem.flatMap { system ⇒
+            provide(ReverseProxyConfig(None, CachedHostConnectionPoolRequestExecutor()(system, mat)))
+          }
         }
       }
     }
   }
 }
 
-case class ProxyUriRequestInterceptor(val uri: Uri) extends ((HttpRequest) ⇒ HttpRequest) {
+private[akka] case class ProxyUriRequestInterceptor(val uri: Uri) extends ((HttpRequest) ⇒ HttpRequest) {
   override def apply(httpRequest: HttpRequest): HttpRequest = {
     httpRequest.withUri(uri)
   }
 }
 
-case object ProxyHeaderRequestInterceptor extends ((HttpRequest) ⇒ HttpRequest) {
+private[akka] case object ProxyHeaderRequestInterceptor extends ((HttpRequest) ⇒ HttpRequest) {
```

Comment:
please also add `/** INTERNAL API */` comment, as `private[akka]` compiles down to `public` for Java (in bytecode)